opm/simulators/linalg/PreconditionerWithUpdate.hpp

```diff
@@ -84,18 +84,49 @@ getDummyUpdateWrapper(Args&&... args)
     return std::make_shared<DummyUpdatePreconditioner<OriginalPreconditioner>>(std::forward<Args>(args)...);
 }
 
-template <class OriginalPreconditioner, class Matrix>
+/// @brief Interface class ensuring make function is overriden
+/// @tparam OriginalPreconditioner - An arbitrary Preconditioner type
+template <class OriginalPreconditioner>
+struct GeneralPreconditionerMaker {
+    virtual std::unique_ptr<
+        Preconditioner<typename OriginalPreconditioner::domain_type, typename OriginalPreconditioner::range_type>>
+    make() = 0;
+};
+
+/// @brief Struct implementing a make function which creates a preconditioner
+/// Can create arbitrary preconditioners using parameter packs as template arguments
+/// @tparam OriginalPreconditioner - An arbitrary preconditioner type
+/// @tparam ...Args - All arguments needed to construct the preconditioner of choice
+template <class OriginalPreconditioner, class... Args>
+struct PreconditionerMaker : public GeneralPreconditionerMaker<OriginalPreconditioner> {
+    PreconditionerMaker(Args&&... args)
+        : args_(args...)
+    {
+    }
+    std::unique_ptr<
+        Preconditioner<typename OriginalPreconditioner::domain_type, typename OriginalPreconditioner::range_type>>
+    make() override
+    {
+        return std::unique_ptr<
```

Comment:
I suggest avoiding the explicit `new` by using `make_unique`:
```
return std::apply([](auto...args){
    return std::make_unique<OriginalPreconditioner>(args...);
}, args_);
```

Reply:
> ```c++
> return std::apply(
>     [](auto...args){
>         return std::make_unique<OriginalPreconditioner>(args...);
>     }, args_);
> ```

Better still:
```C++
return std::apply(
    [](auto&&... args) {
        return std::make_unique<OriginalPreconditioner>(std::forward<Args>(args)...);
    }, args_);
```
There's no need to copy the arguments in this case.

Reply:
I now use the latest suggestion